pkg/messaging/msgtopic.go

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Abstract Machines
+// SPDX-License-Identifier: Apache-2.0
+
+package messaging
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/absmach/supermq/pkg/errors"
+)
+
+const (
+	MsgTopicPrefix     = "m"
+	ChannelTopicPrefix = "c"
+
+	numGroups     = 4 // entire expression + domain group + channel group + subtopic group
+	domainGroup   = 1 // domain group is first in msg topic regexp
+	channelGroup  = 2 // channel group is second in msg topic regexp
+	subtopicGroup = 3 // subtopic group is third in msg topic regexp
+)
+
+var (
+	ErrMalformedTopic    = errors.New("malformed topic")
+	ErrMalformedSubtopic = errors.New("malformed subtopic")
+	// Regex to group topic in format m.<domain_id>.c.<channel_id>.<sub_topic> `^\/?m\/([\w\-]+)\/c\/([\w\-]+)(\/[^?]*)?(\?.*)?$`.
+	msgTopicRegExp = regexp.MustCompile(`^\/?` + MsgTopicPrefix + `\/([\w\-]+)\/` + ChannelTopicPrefix + `\/([\w\-]+)(\/[^?]*)?(\?.*)?$`)
+)
+
+func ParsePublishTopic(topic string) (string, string, string, error) {
```

Comment:
Let's use named return values for clarity `func ParsePublishTopic(topic string) (domainID, chanID, subtopic string, err error)`.